gen/generate.go

```diff
@@ -71,41 +71,87 @@ func Generate(m *compile.Module, o *Options) error {
 			o.OutputDir)
 	}
 
+	importer := thriftPackageImporter{
+		ImportPrefix: o.PackagePrefix,
+		ThriftRoot:   o.ThriftRoot,
+	}
+
 	if o.NoRecurse {
-		return generateModule(m, o)
+		return generateModule(m, importer, o.OutputDir)
 	}
 
 	return m.Walk(func(m *compile.Module) error {
-		if err := generateModule(m, o); err != nil {
+		if err := generateModule(m, importer, o.OutputDir); err != nil {
 			return generateError{Name: m.ThriftPath, Reason: err}
 		}
 		return nil
 	})
 }
 
+type thriftPackageImporter struct {
+	ImportPrefix string
+	ThriftRoot   string
+}
+
+// RelativePackage returns the import path for the top-level package of the
+// given Thrift file relative to the ImportPrefix.
+func (i thriftPackageImporter) RelativePackage(file string) (string, error) {
+	return filepath.Rel(i.ThriftRoot, strings.TrimSuffix(file, ".thrift"))
+}
+
+// Package returns the import path for the top-level package of the given Thrift
+// file.
+func (i thriftPackageImporter) Package(file string) (string, error) {
+	pkg, err := i.RelativePackage(file)
+	if err != nil {
+		return "", err
+	}
+	return filepath.Join(i.ImportPrefix, pkg), nil
+}
+
+// ServicePackage returns the import path for the package for the Thrift service
+// with the given name defined in the given file.
+func (i thriftPackageImporter) ServicePackage(file, name string) (string, error) {
+	topPackage, err := i.Package(file)
+	if err != nil {
+		return "", err
+	}
+
+	return filepath.Join(topPackage, strings.ToLower(name)), nil
```

Comment:
Service names have to be valid identifiers so they won't contain spaces or dashes. Underscores may be present but I think that's okay to have because we want to support `service FooBar` and `service Foo_Bar` defined at the same time. We don't need to support `service foo` and `service Foo` defined at the same time because we validate case insensitivity at compile time.